mlflow/gateway/__init__.py

```diff
@@ -0,0 +1,133 @@
+import logging
+import os
+import pathlib
+import signal
+import sys
+import time
+from typing import Optional
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.constants import CONF_PATH_ENV_VAR, SERVER_HOUSEKEEPING_PAUSE
+from mlflow.protos.databricks_pb2 import BAD_REQUEST
+
+import subprocess
+
+server_process: Optional[subprocess.Popen] = None
+gateway_host: Optional[str] = None
+gateway_port: Optional[int] = None
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.DEBUG)
+_stream_handler = logging.StreamHandler()
+_stream_handler.setLevel(logging.DEBUG)
+formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+_stream_handler.setFormatter(formatter)
+_logger.addHandler(_stream_handler)
+_logger.propagate = False
+
+
+def start_server(config_path: str, host, port):
+    """
+    Starts the server with the given configuration, host and port.
+
+    This function sets the global server_process, gateway_host, and gateway_port variables.
+    It starts the server by creating a new subprocess that runs the server script.
+
+    Args:
+        config_path (str): Path to the configuration file.
+        host (str): The hostname to use for the server.
+        port (int): The port number to use for the server.
+
+    Returns:
+        None
+
+    Raises:
+        subprocess.CalledProcessError: If the server subprocess fails to start.
+    """
+    global server_process
+    global gateway_host
+    global gateway_port
+    os.environ[CONF_PATH_ENV_VAR] = config_path
+
+    if server_process is not None:
+        raise MlflowException(
+            f"There is a currently running server instance at pid '{server_process.pid}'. Please "
+            "terminate the running server instance prior to starting a new instance within this "
+            "context.",
+            error_code=BAD_REQUEST,
+        )
+
+    cmd_path = pathlib.Path(__file__).parent.joinpath("gateway_app.py")
+    server_cmd = [sys.executable, str(cmd_path), "--host", host, "--port", str(port)]
+    gateway_host = host
+    gateway_port = port
+    server_process = subprocess.Popen(server_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    time.sleep(SERVER_HOUSEKEEPING_PAUSE)
```

Comment:
Is sleeping for 2 seconds reliable / do we need it?

Ideally, it would be great to poll in a loop with some timeout - e.g. if the server hasn't started in 30 seconds, throw.

Reply:
exponential backoff polling strategy incoming shortly